src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexRunner.cs

```diff
@@ -207,6 +173,86 @@ public abstract class RegexRunner
+            runtextstart = textstart;
+            runtextbeg = 0;
+            runtextend = text.Length;
+            runtextpos = textstart;
+
+            if (runmatch is null)
+            {
+                // Use a hashtabled Match object if the capture numbers are sparse
+                runmatch = runregex!.caps is null ?
+                    new Match(runregex, runregex.capsize, runtext, runtextbeg, runtextend - runtextbeg, runtextstart) :
+                    new MatchSparse(runregex, runregex.caps, runregex.capsize, runtext, runtextbeg, runtextend - runtextbeg, runtextstart);
+            }
+            else
+            {
+                runmatch.Reset(runregex!, runtext, runtextbeg, runtextend, runtextstart);
+            }
+
+            // Note we test runcrawl, because it is the last one to be allocated
+            // If there is an alloc failure in the middle of the three allocations,
+            // we may still return to reuse this instance, and we want to behave
+            // as if the allocations didn't occur.
+            if (runcrawl != null)
+            {
+                runtrackpos = runtrack!.Length;
+                runstackpos = runstack!.Length;
+                runcrawlpos = runcrawl.Length;
+                return;
+            }
+
+            // Everything above runs once per match.
+            // Everything below runs once per runner.
+
+            InitTrackCount();
+
+            int stacksize;
+            int tracksize = stacksize = runtrackcount * 8;
+
+            if (tracksize < 32)
+            {
+                tracksize = 32;
+            }
+            if (stacksize < 16)
+            {
+                stacksize = 16;
+            }
+
+            runtrack = new int[tracksize];
+            runtrackpos = tracksize;
+
+            runstack = new int[stacksize];
+            runstackpos = stacksize;
+
+            runcrawl = new int[32];
+            runcrawlpos = 32;
+        }
+
+        internal void InitializeTimeout(TimeSpan timeout)
+        {
+            // Handle timeout argument
+            _timeout = -1; // (int)Regex.InfiniteMatchTimeout.TotalMilliseconds
+            bool ignoreTimeout = _ignoreTimeout = Regex.InfiniteMatchTimeout == timeout;
+            if (!ignoreTimeout)
+            {
+                // We are using Environment.TickCount and not Stopwatch for performance reasons.
+                // Environment.TickCount is an int that cycles. We intentionally let timeoutOccursAt
+                // overflow it will still stay ahead of Environment.TickCount for comparisons made
+                // in DoCheckTimeout().
+                Regex.ValidateMatchTimeout(timeout); // validate timeout as this could be called from user code due to being protected
+                _timeout = (int)(timeout.TotalMilliseconds + 0.5); // Round;
+                _timeoutOccursAt = Environment.TickCount + _timeout;
+                _timeoutChecksToSkip = TimeoutCheckFrequency;
```

Comment:
I expect all of this is preventing InitializeTimeout from being inlined.  You might check to see if it is and then whether refactoring it to something more like this helps in the common case of no timeout:
```C#
internal void InitializeTimeout(TimeSpan timeout)
{
    _ignoreTimeout = true;
    if (Regex.InfiniteMatchTimeout != timeout)
    {
        ConfigureTimeout(timeout);

        void ConfigureTimeout(TimeSpan timeout)
        {
            Regex.ValidateMatchTimeout(timeout);
            _ignoreTimeout = false;
            _timeout = (int)(timeout.TotalMilliseconds + 0.5);
            _timeoutOccursAt = Environment.TickCount + _timeout;
            _timeoutChecksToSkip = TimeoutCheckFrequency;
        }
    }
}
```